### PR TITLE
Feature/replace binascii

### PR DESCRIPTION
**Pull request by NotGyro**
### Changes
- Removes any dependencies on the `binascii` crate
- Binascii functions for conversions to and from hexadecimal have been replaced with functions from the `hex` crate 
- Binascii functions for conversions to and from base-64 have been replaced with functions from the `base64` crate 

### Motivation
As stated in Issue #961,
> The hex crate has (for a long while now) been usable within a no-std context, and now also includes an alloc feature. It's also already used in the enclaves in some places, so we should ditch the binascii crate usage in the enclave in favor of hex in order to reduce the number of dependencies we have. 

It's also worth noting that Binascii has not been updated since Feb. 17, 2020.

### Considerations
The test `cyclic_1` is failing on this branch. However, it's also failing on master on my machine. Everything else is passing.

### CODE REVIEW

File: attest/core/src/ias/verify.rs [hunk 1]
```diff
@@ -493,14 +489,11 @@ impl FromHex for VerificationSignature {
...
-            let buffer = b64decode(s.as_bytes(), data.as_mut_slice())?;
-            buffer.len()
-        };
-        data.truncate(buflen);
-        Ok(VerificationSignature::from(data))
+        let result_len = s.len() * 2;
+        let mut result = Vec::with_capacity(result_len);
+        hex::decode_to_slice(s, &mut result)?;
+        result.truncate(result_len);
+        Ok(VerificationSignature::from(result))
```
Comment (iamalwaysuncomfortable) [inline attest/core/src/ias/verify.rs:496]: What exactly is happening right here?

**Review by iamalwaysuncomfortable — APPROVED** [review]
A few questions about some significant seeming changes. Also I think the lint might fail the first time around, but otherwise looks good.

File: docker/install_sgx.sh [hunk 2]
```diff
@@ -24,7 +24,7 @@ cd /tmp
 (
 	. /etc/os-release
 
-	wget "https://download.01.org/intel-sgx/sgx-linux/2.15/distro/ubuntu${VERSION_ID}-server/sgx_linux_x64_sdk_2.15.100.3.bin"
+	wget "https://download.01.org/intel-sgx/sgx-linux/2.15/distro/ubuntu20.04-server/sgx_linux_x64_sdk_2.15.100.3.bin"
```
[the remark this answers is not in the document]
Comment (NotGyro) [reply docker/install_sgx.sh:27]: Apologies, this was a quick hack when I was trying to get SGX working on my machine. This was not intended to go on the repo. I'll revert it.

File: attest/core/src/nonce.rs [hunk 3]
```diff
@@ -139,15 +138,18 @@ impl FromHex for IasNonce {
...
 
 impl ToHex for IasNonce {
     fn to_hex(&self, dest: &mut [u8]) -> Result<usize, usize> {
-        match bin2hex(self.as_ref(), dest) {
-            Ok(buffer) => Ok(buffer.len()),
+        let bytes = self.as_ref();
+        match hex::encode_to_slice(bytes, dest) {
+            // Previously binascii was used instead of hex. 
+            // Binascii would always return input.len() * 2.
+            Ok(()) => Ok(bytes.len()*2),
```
Comment (iamalwaysuncomfortable) [inline attest/core/src/nonce.rs:152]: Is it still needed to return input.len() * 2

File: crypto/keys/src/traits.rs [hunk 4]
```diff
@@ -74,11 +74,10 @@ impl<T: PublicKey + DistinguishedEncoding> Fingerprintable for T {
         let output_size = D::OutputSize::to_usize();
         let mut hash_strbuf: Vec<u8> = vec![0u8; output_size * 2];
         let hash_len = hash_strbuf.len();
-        let hash_len = {
-            let hash_slice = binascii::bin2hex(&hash, &mut hash_strbuf)
-                .map_err(|_e| KeyError::LengthMismatch(hash_len, output_size * 2))?;
-            hash_slice.len()
-        };
+        
+        hex::encode_to_slice(&hash, &mut hash_strbuf)
```
Comment (remoun) [inline crypto/keys/src/traits.rs:78]: IIUC this impl can be reduced to:

```rust
let hex_str = hex::encode(&hash).map_err(|_e| KeyError::LengthMismatch(hash_len, output_size * 2))?;
// Safety: hex::encode() returns ASCII chars, so each char should be one byte.
let pairs = hex_str.as_bytes().chunks(2).map(std::str::from_utf8).collect::<Result<Vec<&str>, _>>().unwrap();
let retval = pairs.join(":");
```

File: attest/core/src/error.rs [hunk 5]
```diff
@@ -239,9 +239,9 @@ pub enum QuoteError {
     InvalidUtf8,
 }
 
-impl From<ConvertError> for QuoteError {
-    fn from(src: ConvertError) -> Self {
-        QuoteError::Encoding(src.into())
+impl From<base64::DecodeError> for QuoteError{ 
+    fn from(_src: base64::DecodeError) -> Self { 
+        QuoteError::Encoding(EncodingError::InvalidInput)
```
Comment (remoun) [inline attest/core/src/error.rs:244]: There's an `InvalidLength` variant of `base64::DecodeError`, so we should map that to the corresponding `EncodingError` variant
```suggestion
impl From<base64::DecodeError> for QuoteError {
    fn from(src: base64::DecodeError) -> Self {
        use base64::DecodeError::*;
        let kind = match src {
            InvalidLength => EncodingError::InvalidInputLength,
            InvalidByte(_, _)|InvalidLastSymbol(_, _) => EncodingError::InvalidInput,
        };
        QuoteError::Encoding(kind)
```

File: fog/report/cli/src/main.rs [hunk 6]
```diff
@@ -272,8 +271,9 @@ fn main() {
     };
 
     let mut hex_buf = [0u8; 64];
-    bin2hex(
-        CompressedRistrettoPublic::from(&pubkey).as_ref(),
+    let compressed_ristretto_public = CompressedRistrettoPublic::from(&pubkey);
+    hex::encode_to_slice(
+        <CompressedRistrettoPublic as AsRef<[u8;32]>>::as_ref(&compressed_ristretto_public),
         &mut hex_buf[..],
     )
```
Comment (remoun) [inline fog/report/cli/src/main.rs:278]: Optional/DRY: `let hex_buf = hex::encode(CompressedRistrettoPublic::from(&pubkey).as_bytes()).expect(...);`

File: test-vectors/vectors/account_keys/acct_priv_keys_from_bip39.jsonl [hunk 7]
```diff
@@ -1,48 +0,0 @@
-{"entropy":[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],"mnemonic":"abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon about","account_index":0,"view_private_key":[188,129,163,113,13,117,198,157,198,39,182,192,73,120,6,53,11,179,210,123,59,88,209,194,153,223,0,109,169,0,192,4],"spend_private_key":[148,183,217,189,73,62,87,252,199,231,25,152,96,19,26,14,154,116,76,19,157,217,127,215,83,252,120,115,140,122,163,6]}
```
Comment (remoun) [inline test-vectors/vectors/account_keys/acct_priv_keys_from_bip39.jsonl:1]: Please revert this file deletion

File: util/encodings/Cargo.toml [hunk 8]
```diff
@@ -9,11 +9,10 @@ description = "Support for various simple encodings (hex strings, base64 strings
...
 base64 = { version = "0.13", default-features = false }
-binascii = "0.1.4"
 displaydoc = { version = "0.2", default-features = false }
 hex = { version = "0.4", default-features = false }
 serde = { version = "1.0", default-features = false, features = ["alloc", "derive"] }
 
 [features]
 default = []
-std = []
+std = []
```
Comment (remoun) [inline util/encodings/Cargo.toml:18]: Nit: Add newline
```suggestion
std = []

```

File: attest/core/src/nonce.rs (hunk 3 above)
Comment (remoun) [reply attest/core/src/nonce.rs:152]: Yes, the `ToHex` trait requires it

File: attest/core/src/types/pib.rs [hunk 9]
```diff
@@ -33,21 +30,16 @@ impl FromBase64 for PlatformInfoBlob {
             return Err(EncodingError::InvalidInputLength);
         }
 
-        let expected_len = src.len() / 4 * 3;
+        let expected_len = base64_buffer_size(src.len());
```
Comment (remoun) [inline attest/core/src/types/pib.rs:33]: This is the inverse calculation; let's leave it as-is for now
```suggestion
        let expected_len = src.len() / 4 * 3;
```

File: attest/core/src/types/pib.rs [hunk 10]
```diff
@@ -33,21 +30,16 @@ impl FromBase64 for PlatformInfoBlob {
...
-        let mut data = vec![0u8; src.len() / 4 * 3];
-        let data_len = {
-            let output = b64decode(src.as_bytes(), data.as_mut_slice())?;
-            output.len()
-        };
-        data.truncate(data_len);
-        let mut retval = Self::default();
-        retval.0.platform_info[..].copy_from_slice(&data[4..(SGX_PLATFORM_INFO_SIZE + 4)]);
-        Ok(retval)
+        let data = base64::encode_config(src.as_bytes(), B64_CONFIG);
```
Comment (remoun) [inline attest/core/src/types/pib.rs:38]:
```suggestion
        let data = base64::decode_config(src.as_bytes(), B64_CONFIG)?;
```

File: crypto/keys/src/traits.rs [hunk 11]
```diff
@@ -6,10 +6,8 @@ pub use digest::Digest;
 pub use ed25519::signature::{DigestSigner, DigestVerifier, Signature, Signer, Verifier};
 pub use mc_util_repr_bytes::{typenum::Unsigned, GenericArray, LengthMismatch, ReprBytes};
 
-// Macros with names that overlap a module name...
-use alloc::vec;
-
 use alloc::{string::String, vec::Vec};
+
```
Comment (remoun) [inline crypto/keys/src/traits.rs:10]: Nit: Unnecessary blank line
```suggestion
```

File: attest/core/src/error.rs [hunk 12]
```diff
@@ -239,9 +239,14 @@ pub enum QuoteError {
...
-    fn from(src: ConvertError) -> Self {
-        QuoteError::Encoding(src.into())
+impl From<base64::DecodeError> for QuoteError {
+    fn from(src: base64::DecodeError) -> Self {
+        use base64::DecodeError::*;
+        let kind = match src {
+            InvalidLength => EncodingError::InvalidInputLength,
+            InvalidByte(_, _)|InvalidLastSymbol(_, _) => EncodingError::InvalidInput,
+        };
+        QuoteError::Encoding(kind)
```
Comment (remoun) [inline attest/core/src/error.rs:249]: I notice EncodingError already implements `From<base64::DecodeError>` so we can further simplify this
```suggestion
        QuoteError::Encoding(src.into())
```

File: attest/core/src/traits.rs [hunk 13]
```diff
@@ -7,14 +7,21 @@ pub(crate) use alloc::format as _alloc_format;
...
     fmt::{Debug, Display, Formatter, Result as FmtResult},
     hash::{Hash, Hasher},
 };
+
+pub(crate) use base64::decode_config_slice as b64decode;
+pub(crate) use base64::encode_config_slice as b64encode;
+pub(crate) use crate::B64_CONFIG;
+
+pub(crate) use hex::encode_to_slice as bin2hex;
+pub(crate) use hex::decode_to_slice as hex2bin;
```
Comment (remoun) [inline attest/core/src/traits.rs:22]: Nit: While these aliases match the old impl, they're not great names. How about `hex_encode`/`hex_decode`?

File: attest/core/src/nonce.rs (hunk 3 above)
Comment (iamalwaysuncomfortable) [reply attest/core/src/nonce.rs]: Helpful to know!

**Review by remoun — APPROVED** [review]

File: fog/report/cli/Cargo.toml [hunk 14]
```diff
@@ -24,6 +24,6 @@ mc-util-keyfile = { path = "../../../util/keyfile" }
 mc-util-uri = { path = "../../../util/uri" }
 
 base64 = "0.13"
-binascii = "0.1.4"
+hex = "0.4"
 clap = { version = "3.1", features = ["derive", "env"] }
 grpcio = "0.10.2"
```
Comment (remoun) [inline fog/report/cli/Cargo.toml:29]: Sort deps to fix the lint error.

```suggestion
clap = { version = "3.1", features = ["derive", "env"] }
grpcio = "0.10.2"
hex = "0.4"
```

**Review by remoun — APPROVED** [review]

**Pull request merged**